### PR TITLE
support entity/entity-alias write, update, and delete operations

### PR DESCRIPTION
**Pull request by dwelch0**
Adds functionality to create, update, and delete entities and entity aliases, based off of data/teams/user files. Entities are created for each user definition that contains a reference to a role which contains permissions for vault. Current implementation only supports creating entities/aliases for users with vault permission. The corresponding entity aliases are of type oidc.

Note: addresses https://github.com/app-sre/vault-manager/pull/35#pullrequestreview-920554832

TICKETS:

* https://issues.redhat.com/browse/APPSRE-4656
* https://issues.redhat.com/browse/APPSRE-4657

### CODE REVIEW

File: tests/app-interface/data.json [hunk 1]
```diff
@@ -11558,6 +11695,56 @@
...
+                },
+                {
+                    "name": "priority",
+                    "type": "string",
+                    "isRequired": true
+                }
+            ]
+        },
+        {
+            "name": "JiraSeverityPriorityMappings_v1",
```
Comment (TGPSKI) [inline tests/app-interface/data.json:11714]: assuming this is an indirect requirement?

File: tests/app-interface/data/dependencies/vault/permissions/oidc/app-sre-vault-admin.yml [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+$schema: /access/oidc-permission-1.yml
+
+labels:
+  service: vault.local
+
+name: app-sre-vault-admin
+service: vault
+description: app-sre vault administrator permission
+vault_policies:
+- $ref: /services/vault/config/policies/app-sre-vault-admin.yml
```
Comment (TGPSKI) [inline tests/app-interface/data/dependencies/vault/permissions/oidc/app-sre-vault-admin.yml:11]: newline

File: tests/fixtures/entities/enable_vault_entities_and_aliases.graphql [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+        service
+        ... on OidcPermissionVault_v1 {
+          vault_policies {
+            name
+          }
+        }
+      }
+    }
+  }
+}
```
Comment (TGPSKI) [inline tests/fixtures/entities/enable_vault_entities_and_aliases.graphql:19]: newline

File: toplevel/identity/entity.go [hunk 4]
```diff
@@ -0,0 +1,642 @@
...
+type entityAlias struct {
+	Name           string
+	Id             string
+	Type           string
+	AuthType       string
+	AccessorId     string
+	CustomMetadata interface{}
+}
+
+var _ vault.Item = entity{}
```
Comment (TGPSKI) [inline toplevel/identity/entity.go:63]: what is the purpose of this?

File: toplevel/identity/entity.go [hunk 5]
```diff
@@ -0,0 +1,642 @@
...
+func (e entity) KeyForDescription() string {
+	return fmt.Sprintf("%v", e.Metadata)
+}
+
+func (e entity) Equals(i interface{}) bool {
+	entry, ok := i.(entity)
+	if !ok {
+		return false
+	}
+	return e.Name == entry.Name &&
```
Comment (TGPSKI) [inline toplevel/identity/entity.go:84]: this looks very nice

File: toplevel/identity/entity.go [hunk 6]
```diff
@@ -0,0 +1,642 @@
...
+		reflect.DeepEqual(e.Policies, entry.Policies)
+}
+
+func (e entity) CreateOrUpdate(action string) {
+	path := filepath.Join("identity", e.Type, "name", e.Name)
+	config := map[string]interface{}{
+		"policies": e.Policies,
+		"metadata": e.Metadata,
+	}
+	vault.WriteSecret(path, config)
```
Comment (TGPSKI) [inline toplevel/identity/entity.go:95]: Assuming there is no tune method available for this mount type?

File: toplevel/identity/entity.go (hunk 4 above)
Comment (TGPSKI) [reply toplevel/identity/entity.go:63]: not saying its wrong I just don't know this pattern immediately off the top of my head

File: toplevel/identity/entity.go [hunk 7]
```diff
@@ -0,0 +1,642 @@
...
+}
+
+func (ea entityAlias) Create(entityId string, customMetadataSupported bool) {
+	path := filepath.Join("identity", ea.Type)
+	config := map[string]interface{}{
+		"name":           ea.Name,
+		"canonical_id":   entityId,
+		"mount_accessor": ea.AccessorId,
+	}
+	if customMetadataSupported {
```
Comment (TGPSKI) [inline toplevel/identity/entity.go:136]: is there a use case for customMetadata?

File: toplevel/identity/entity.go [hunk 8]
```diff
@@ -0,0 +1,642 @@
...
+		config["custom_metadata"] = ea.CustomMetadata
+	}
+	vault.WriteEntityAlias(path, config)
+	log.WithField("path", filepath.Join(path, ea.Name)).WithField("type", ea.AuthType).Info(
+		"[Vault Identity] entity alias successfully written")
+}
+
+func (ea entityAlias) Update(entityId string, customMetadataSupported bool) {
```
Comment (TGPSKI) [inline toplevel/identity/entity.go:144]: clearly entities don't have update and aliases do, makes sense

File: toplevel/identity/entity.go [hunk 9]
```diff
@@ -0,0 +1,642 @@
...
+	}
+	if customMetadataSupported {
+		config["custom_metadata"] = ea.CustomMetadata
+	}
+	vault.WriteSecret(path, config)
+	log.WithField("path", filepath.Join(path, ea.Name)).WithField("type", ea.AuthType).Info(
+		"[Vault Identity] entity alias successfully updated")
+}
+
+func (ea entityAlias) Delete() {
```
Comment (TGPSKI) [inline toplevel/identity/entity.go:159]: EA

ITS IN THE GAME

File: toplevel/identity/entity.go [hunk 10]
```diff
@@ -0,0 +1,642 @@
...
+											},
+										},
+									},
+									Metadata: map[string]interface{}{
+										"name": entry.Name,
+									},
+								}
+								uniquePolicies = make(map[string]bool)
+							}
+							for _, policy := range permission.Policies {
```
Comment (TGPSKI) [inline toplevel/identity/entity.go:271]: I'm curious why policies are managed in this block. Isn't there another part of vault-manager which handles policies?

Comment (TGPSKI) [reply toplevel/identity/entity.go:271]: Does this have the side effect of writing policies?

Comment (TGPSKI) [reply toplevel/identity/entity.go:271]: If policies are included just to create entities and entity aliases for users who have roles, this is OK. If we are also managing and writing policies in this block, we should re-work the approach.

File: toplevel/identity/entity.go [hunk 11]
```diff
@@ -0,0 +1,642 @@
...
+			Aliases: processedAliases,
+		})
+	}
+	return processed, nil
+}
+
+// performs concurrent requests to retrieve additional details for existing entities/entity aliases
+// these details require explicit requests to vault api for each entitiy/alias
+func getExistingEntitiesDetails(entities []entity, threadPoolSize int, customMetadataSupported bool) {
+	bwg := utils.NewBoundedWaitGroup(threadPoolSize)
```
Comment (TGPSKI) [inline toplevel/identity/entity.go:371]: A+

File: toplevel/identity/entity.go [hunk 12]
```diff
@@ -0,0 +1,642 @@
...
+						for k, v := range rawAlias["custom_metadata"].(map[string]interface{}) {
+							if ok {
+								aliasMap[k] = v
+							}
+						}
+					}
+				}
+			}
+
+			entity.Policies = policies
```
Comment (TGPSKI) [inline toplevel/identity/entity.go:417]: Let's review the usage of policies on a review call on Monday. I'm sure this is OK, want to hear the flow from you.

File: toplevel/identity/entity.go (hunk 11 above)
Comment (TGPSKI) [reply toplevel/identity/entity.go:371]: Love the goroutine usage here

File: toplevel/identity/entity.go [hunk 13]
```diff
@@ -0,0 +1,642 @@
...
+	var accessorIds map[string]string
+	// extra work (vault api request) required to organize accessor ids
+	if len(aliasesToBeWritten) > 0 {
+		accessorIds = make(map[string]string)
+		authBackends := vault.ListAuthBackends()
+		for k, v := range authBackends {
+			accessorIds[strings.TrimRight(k, "/")] = v.Accessor
+		}
+	}
+	if _, exists := aliasesToBeWritten["id"]; exists {
```
Comment (TGPSKI) [inline toplevel/identity/entity.go:483]: I want to understand the thought on this block and the next in person

**Review by TGPSKI — COMMENTED** [review]
It looks good overall. Let's video on Monday and get this locked in.

File: toplevel/identity/entity.go (hunk 4 above)
Comment (dwelch0) [reply toplevel/identity/entity.go:63]: it was a new pattern to me as well but a practice employed by the other integrations so I opted to include. It is a "check" that occurs during compile time to see if the `entity` structure implements the `vault.Item` interface correctly.

File: toplevel/identity/entity.go (hunk 5 above)
Comment (dwelch0) [reply toplevel/identity/entity.go:84]: :bow:

File: toplevel/identity/entity.go (hunk 6 above)
Comment (dwelch0) [reply toplevel/identity/entity.go:95]: Correct. Both create and update use the same method and parameters in vault api, in which the secret is simply written to the desired path.

File: toplevel/identity/entity.go (hunk 7 above)
Comment (dwelch0) [reply toplevel/identity/entity.go:136]: Unsure. I opted to write the logic to include it as it seemed easier to remove after the fact than add it in later. Maybe for trivial info like which team someone is apart of etc. Can easily rip out if need be.

File: toplevel/identity/entity.go (hunk 13 above)
Comment (dwelch0) [reply toplevel/identity/entity.go:483]: Understood. A call will be worthwhile. The logic for aliases became fairly more complex than that for entities due to alias dependency on entity creation beforehand.

File: query.graphql [hunk 14]
```diff
@@ -102,4 +102,21 @@
...
+        service
+        ... on OidcPermissionVault_v1 {
+          vault_policies {
+            name
+          }
+        }
+      }
+    }
+  }
+}
```
Comment (bkez322) [inline query.graphql:122]:
```suggestion
}

```

**Comment by bkez322** [comment]
This is really fantastic.

**Pull request merged**